### ISSUE TITLE
Roll our own replacement for NINT

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#441]
Surprisingly, gfortran uses `lround` to compute the value of `NINT` and this is relatively expensive: about 1.6% of our total run time is spent calling this function. Even more surprising: this is apparently more expensive than `SQRT` since gfortran manages to inline the later but not the former.

Since this is called when we already know that the resulting integer is positive (I think) we could just roll our own 'round up' function that we can compel the fortran compiler to always inline. This would be a relatively simple way to boost performance in the IB kernels (they spend about 7.5% percent of their time calling this function).

**Comment by drwells** [comment]
`-finline-functions` might be enough to make this happen; I am investigating.

**Comment by drwells** [comment]
Nope: turning on more optimizations doesn't seem to help at all (the number of cycles is identical).

**Comment by boyceg** [comment]
Intel docs ([see here](https://software.intel.com/en-us/fortran-compiler-developer-guide-and-reference-nint)) suggest implementing `NINT(a)` as `INT(a+0.5)` for `a .gt. 0` and as `INT(a-0.5)` for `a .le. 0`.

**Comment by drwells** [comment]
I am a bit annoyed that gfortran, with maximum optimizations, cannot inline this function :(

**Comment by boyceg** [comment]
I just started down the rabbit hole of compiling the FORTRAN kernels with `-fopt-info-all`... Some inner loops are getting vectorized.

**Comment by boyceg** [comment]
One annoying thing here is that, because I haven't managed to escape FORTRAN 77 yet, changing `NINT` to a new function with a longer name requires re-wrapping a bunch (too many!) lines.

**Comment by boyceg** [comment]
Apparently there is a link-time optimization flag, `-flto`, that may result in more aggressive inlining with gfortran.

**Comment by boyceg** [comment]
You could try `flang` (LLVM Fortran compiler). Apparently you can get it through Spack.

**Comment by boyceg** [comment]
> Apparently there is a link-time optimization flag, `-flto`, that may result in more aggressive inlining with gfortran.

(If you mix-and-match LLVM and GCC compilers like I do on MacOS, this probably will not be effective unless you can use the GCC compilers for linking.)

**Comment by drwells** [comment]
I have no idea if `-flto` can be used with multiple compilers; IIRC GCC actually outputs some GIMPL and other intermediate data structures that it uses to do more code generation at the end.
